api/src/wfl/module/arrays.clj

```diff
@@ -0,0 +1,224 @@
+  "The WDL label applied to Cromwell metadata."
+  {(keyword wfl/the-name)
+   pipeline})
+
+(def cromwell-label
+  "The WDL label applied to Cromwell metadata."
+  (let [[key value] (first cromwell-label-map)]
+    (str (name key) ":" value)))
+
+(def fingerprinting
+  "Fingerprinting inputs for arrays."
+  {:haplotype_database_file              "gs://gcp-public-data--broad-references/hg19/v0/Homo_sapiens_assembly19.haplotype_database.txt"
+   :variant_rsids_file                   "gs://broad-references-private/hg19/v0/Homo_sapiens_assembly19.haplotype_database.snps.list"})
+
+(def other-inputs
+  "Miscellaneous inputs for arrays."
+  {:contamination_controls_vcf       nil
+   :subsampled_metrics_interval_list nil
+   :disk_size                        100
+   :preemptible_tries                3})
+
+(defn map-arrays-environment
+  "Map ARRAYS-ENV to environment for inputs preparation."
+  [arrays-env]
+  ({:arrays-dev "dev" :arrays-prod "prod"} arrays-env))
+
+(defn env-inputs
+  "Array inputs for ENVIRONMENT that do not depend on the input file."
+  [environment]
+  {:vault_token_path (get-in env/stuff [environment :vault_token_path])
+   :environment      (map-arrays-environment environment)})
+
+(defn get-per-sample-inputs
+  "Throw or return per-sample INPUTS."
+  [inputs]
+  (let [mandatory-keys [:analysis_version_number
+                        :bead_pool_manifest_file
+                        :call_rate_threshold
+                        :chip_well_barcode
+                        :cluster_file
+                        :extended_chip_manifest_file
+                        :fingerprint_genotypes_vcf_file
+                        :fingerprint_genotypes_vcf_index_file
+                        :green_idat_cloud_path
+                        :params_file
+                        :red_idat_cloud_path
+                        :reported_gender
+                        :sample_alias
+                        :sample_lsid]
+        optional-keys  [;; genotype concordance inputs
+                        :control_sample_vcf_file
+                        :control_sample_vcf_index_file
+                        :control_sample_intervals_file
+                        :control_sample_name
+                        :genotype_concordance_threshold
+                        ;; cloud path of a thresholds file to be used with zCall
+                        :zcall_thresholds_file
+                        ;; cloud path of the Illumina gender cluster file
+                        :gender_cluster_file
+                        ;; arbitrary path to be used by BAFRegress
+                        :minor_allele_frequency_file]
+        mandatory      (select-keys inputs mandatory-keys)
+        optional       (select-keys inputs optional-keys)
+        missing        (vec (keep (fn [k] (when (nil? (k mandatory)) k)) mandatory-keys))]
+    (when (seq missing)
+      (throw (Exception. (format "Missing per-sample inputs: %s" missing))))
+    (merge optional mandatory)))
```

Comment:
We might just want to return `inputs` here.

Reply:
If it's okay with you @tbl3rd I'd rather change this behavior in a separate PR that includes the aou module as well